src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java

```diff
@@ -20,45 +20,171 @@
-import sparksoniq.spark.iterator.flowr.base.FlowrClauseSparkIterator;
 
 import java.util.ArrayList;
 import java.util.List;
 
-public class ForClauseSparkIterator extends FlowrClauseSparkIterator {
+public class ForClauseSparkIterator extends SparkRuntimeTupleIterator {
+
+    private String _variableName;           // for efficient use in local iteration
+    private RuntimeIterator _expression;
+    private DynamicContext _tupleContext;   // re-use same DynamicContext object for efficiency
+    private FlworTuple _nextLocalTupleResult;
+    private FlworTuple _inputTuple;     // tuple received from child, used for tuple creation
 
     public ForClauseSparkIterator(RuntimeTupleIterator child, VariableReferenceIterator variableReference,
                                   RuntimeIterator assignmentExpression, IteratorMetadata iteratorMetadata) {
-        super(child, null, FLWOR_CLAUSES.FOR, iteratorMetadata);
-        this._children.add(variableReference);
-        this._children.add(assignmentExpression);
+        super(child, iteratorMetadata);
+        _variableName = variableReference.getVariableName();
+        _expression = assignmentExpression;
+    }
+
+    @Override
+    public boolean isRDD() {
+        return (_expression.isRDD() || (_child != null && _child.isRDD()));
+    }
+
+    @Override
+    public FlworTuple next() {
+        if(_hasNext == true){
+            FlworTuple result = _nextLocalTupleResult;      // save the result to be returned
+            // calculate and store the next result
+            if (_child == null) {       // if it's the initial for clause, call the correct function
+                initialFor_setNextLocalTupleResult();
+            } else {
+                setNextLocalTupleResult();
+            }
+            return result;
+        }
+        throw new IteratorFlowException("Invalid next() call in let flwor clause", getMetadata());
+    }
+
+    private void setNextLocalTupleResult() {
+        if (_expression.isOpen()) {
+            boolean isResultSet = setResultFromExpression();
+            if (isResultSet) {
+                return;
+            }
+        }
+
+        while (_child.hasNext()) {
+            _inputTuple = _child.next();
+            _tupleContext.removeAllVariables();             // clear the previous variables
+            _tupleContext.setBindingsFromTuple(_inputTuple);      // assign new variables from new tuple
+
+            _expression.open(_tupleContext);
+            boolean isResultSet = setResultFromExpression();
+            if (isResultSet) {
+                return;
+            }
+        }
+
+        // execution reaches here when there are no more results
+        _child.close();
+        this._hasNext = false;
+    }
+
+    /**
+     * _expression has to be open prior to call.
+     * @return true if _nextLocalTupleResult is set and _hasNext is true, false otherwise
+     */
+    private boolean setResultFromExpression() {
```

Comment:
You can merge these two functions in just one that takes a boolean. This avoids code duplication.